Fix tests failing on WSL systems.

rsashank opened the pull request:
### What does this PR do, and why?

Addressed issues with file path separators `\\` and GUI return codes specific to WSL environments.

Fixes the failing of these tests on WSL environments, these have been failing for a while now but I just got used to ignoring them, decided to look into them today and fix. 😓 

![image](https://github.com/zulip/zulip-terminal/assets/110327345/4c672e63-2416-4df1-80b2-f876d626fc14)

<!-- See README for documentation, or ask in #zulip-terminal if unclear -->

### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->
- [ ] 

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in [WSL specific test fails](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/WSL.20specific.20test.20fails/near/1817087)
- [ ] Fully fixes #
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [ ] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [ ] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [ ] It has a commit summary describing the  motivation and reasoning for the change
- [ ] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->

rsashank committed:
refactor: helper: Fix WSL normalized_file_paths related tests.

Addressed issues with file path separators `\\` specific to WSL
environments.

rsashank commented on tests/helper/test_helper.py:
```diff
@@ -539,6 +540,9 @@ def test_download_media(
     ) = media_path
     controller = mocker.Mock()
 
+    if PLATFORM == "WSL":
+        media_path = media_path.replace("/", "\\")
+
```
I'm not sure if I can import PLATFORM into tests, but this seemed like the only way to address the difference in file paths in WSL.

rsashank replied:
To further clarify, based on a discussion on CZO, the issue stems from the `normalized_path` function used in the `download_media` method. The `media_path` assertion doesn't account for the WSL-specific normalized path.

neiljp replied:
For a conditional in a test, the cleaner way to handle this would normally be to parametrize the test (over PLATFORM), though note you'd want a third path for the `expected_return_value` or similar output parameter, since it doesn't correspond to `url` or `media_path` in every case (though the latter may not be representative - see below re confirming).

The alternative to this conditional/parametrizing solution, which may also be clearer in the long term, is related to how (as we confirmed in the channel) the path returned is the path used for *opening* the temporary file, not the path used for *saving* it. So it would be clearer to refactor the code before fixing the tests, such that the `normalized_file_path` occurs outside of this function, and this function would be platform-independent to test.

rsashank replied:
Moved the `normalized_file_path` to be called in `process_media` 👍 
 Clear explanation, thanks!

rsashank commented on tests/helper/test_helper.py:
```diff
@@ -604,14 +608,14 @@ def test_process_media_empty_url(
 @pytest.mark.parametrize(
     "returncode, error",
     [
-        (0, []),
+        (1 if PLATFORM == "WSL" else 0, []),
         (
-            1,
+            0 if PLATFORM == "WSL" else 1,
             [
                 " The tool ",
                 ("footer_contrast", "xdg-open"),
                 " did not run successfully" ". Exited with ",
-                ("footer_contrast", "1"),
+                ("footer_contrast", "0" if PLATFORM == "WSL" else "1"),
             ],
         ),
     ],
```
This test didn't consider the different return codes in WSL, as given in `platform_code.py` line 92.

```python
def successful_GUI_return_code() -> int:  # noqa: N802 (allow upper case)
    """
    Returns success return code for GUI commands, which are OS specific.
    """
    # WSL uses GUI return code as 1. Refer below link to know more:
    # https://stackoverflow.com/questions/52423031/
    # why-does-opening-an-explorer-window-and-selecting-a-file-through-pythons-subpro/
    # 52423798#52423798
    if PLATFORM == "WSL":
        return 1

    return 0```

rsashank replied:
`explorer.exe` seems to always return a non-zero exit code, regardless of whether the task is successful.

More information could be found on this [issue](https://github.com/microsoft/WSL/issues/6565).

rsashank replied:
> we can refactor the function you quoted, such that it returns some value, depending upon a passed-in return code. This would move the comparison, and hence these explicit error codes, into a platform-dependent location.

Implemented this approach, thanks for the suggestion! Also will look into refactoring some `open_media` logic to a platform-specific location in another PR.

Niloth-p commented:
Nice work! Thanks for addressing those.
The motivation and code LGTM.
As discussed, this is not something that I can run and test though, as I don't use WSL.

neiljp commented:
@rsashank I started a review, but will leave that pending until I hear back to confirm the implementation details I noted in #zulip-terminal.

neiljp reviewed (commented):
@rsashank Thanks for finding these problems and working on solutions - we don't have WSL in CI (GitHub PR tests), so some of these issues are not always obvious, and not everything can be tested in CI in any case.

Following the discussion in the channel, I've adjusted some comments, though the same general gist of my suggestions remains the same, ie. that refactoring slightly first would be preferable, and may allow further subsequent refactoring which would make everything a lot easier to read.

Due to limited resources for testing on each platform, in general its better if we can isolate as much into platform-code as we can, during times while people are available to test it manually. That applies to this code in `helper`, as well as anywhere else we explicitly rely upon PLATFORM. That allows the platform-dependent code to be manually-tested well, and the interface between that and the application to be as clean as possible, as well as tests like these to be less fragile :)

rsashank committed:
bugfix: tests: Fix WSL GUI exit status related tests.

WSL always returns a non-zero exit code. The `successful_GUI_return_code()`
function has been refactored to directly compare exit statuses and return
"success" or "failure" accordingly.

Fix open_media test to validate both the tool and the error code across
various platforms.

Refactor process_media to shift platform dependent code to platform_code.py.

neiljp commented on tests/platform_code/test_platform_code.py:
```diff
@@ -76,20 +76,31 @@ def test_notify_quotes(
+        ("WSL", 1, "success"),
+        ("Linux", 1, "failure"),
+        ("MacOS", 1, "failure"),
+        # NOTE: WSL always returns a non-zero exit code (1), so we do not test for it.
```
As per another comment, we should test for this, even if it results in the same response - we're testing the function, not the system behavior, in this test.

rsashank replied:
I removed the failure handling entirely since it's considered an abnormality that might be addressed in the future. `explorer.exe` seems to be intended to return 1 for failure and 0 for success, so returning 0 for a failure didn't make sense. 

This could be updated if WSL fixes this, what do you think?

neiljp reviewed (commented):
@rsashank I hoped to merge this fully but ended up leaving a few comments aimed primarily at the second commit - I've picked off the first commit and merged it already, after making the change I indicated in the comment re which item to patch.

Hopefully that first commit reduces the failing tests you're finding regularly, and the other commit will be quick to follow :)

rsashank commented on tests/helper/test_helper.py:
```diff
@@ -573,6 +573,7 @@ def test_process_media(
     )
     mocked_open_media = mocker.patch(MODULE + ".open_media")
     mocker.patch(MODULE + ".PLATFORM", platform)
+    mocker.patch("zulipterminal.platform_code.PLATFORM", platform)
```
While testing locally in my WSL environment, the platform detection incorrectly identifies the platform as WSL when either one of the patches is missing. This causes the tests to enter the if PLATFORM == 'WSL' condition and change the path.

So, I had to mock the PLATFORM value in both the platform_code and helper.

I'm not entirely sure why this is happening, but patching both seems to ensure the consistency for all platforms that are being tested.
![image](https://github.com/user-attachments/assets/4022caa0-d7cf-43b0-a4ca-c518d3ffa0f3)

neiljp replied:
Thanks for this feedback! This seems likely since we still use PLATFORM in both locations for this :/

This appears to be the only outstanding change in this commit after a rebase, since I merged the first commit previously, so the commit text is currently inaccurate at that point. It'd also be useful to group and comment the two lines, if they appear to be necessary to pass the tests, as well as update the commit.

This is a 'test bugfix' commit, specific to WSL, essentially?

rsashank replied:
Grouped them together and added a comment, also changed commit message.

Yes, it's a test bugfix specific to WSL.

rsashank commented on zulipterminal/platform_code.py:
```diff
@@ -89,18 +89,19 @@ def notify(title: str, text: str) -> str:
-    # why-does-opening-an-explorer-window-and-selecting-a-file-through-pythons-subpro/
-    # 52423798#52423798
+    # NOTE: WSL always returns a non-zero exit code (1) for GUI commands.
+    # This is a known issue. Therefore, we consider it a success if the exit code is 1.
+    # For more information, see: https://github.com/microsoft/WSL/issues/6565
     if PLATFORM == "WSL":
-        return 1
+        return "success" if exit_status == 1 else "failure"
```
Thank you for pointing this out. I've removed the mention of "failure" handling.

The linked issue specifically refers to explorer.exe, and since validate_GUI_exit_status only deals with explorer.exe, I believe it should be fine to document here. What do you say?

rsashank commented on zulipterminal/platform_code.py:
```diff
@@ -89,18 +89,19 @@ def notify(title: str, text: str) -> str:
 
 
-def successful_GUI_return_code() -> int:  # noqa: N802 (allow upper case)
+def validate_GUI_exit_status(  # noqa: N802 (allow upper case)
+    exit_status: int,
+) -> str:
```
Got it!

rsashank commented on tests/platform_code/test_platform_code.py:
```diff
@@ -76,20 +76,31 @@ def test_notify_quotes(
+        ("MacOS", 1, "failure"),
+        # NOTE: WSL always returns a non-zero exit code (1), so we do not test for it.
+    ],
+    ids=[
+        "Linux_0_return_code",
+        "MacOS_0_return_code",
+        "WSL_1_return_code",
+        "Linux_1_return_code",
+        "MacOS_1_return_code",
     ],
```
Got it.

rsashank commented on tests/helper/test_helper.py:
```diff
@@ -603,28 +603,53 @@ def test_process_media_empty_url(
     media_path: str = "/tmp/zt-somerandomtext-image.png",
 ) -> None:
     mocked_run = mocker.patch(MODULE + ".subprocess.run")
+    mocker.patch("zulipterminal.platform_code.PLATFORM", platform)
```
Doesn't seem to work on my local WSL environment - trying to figure out why :) 

I need to have both patched.

rsashank commented on tests/helper/test_helper.py:
```diff
@@ -603,28 +603,53 @@ def test_process_media_empty_url(
+                ("footer_contrast", "1"),
+            ],
+        ),
+        # NOTE: WSL always returns a non-zero exit code (1), so we do not test for it.
```
True, explorer.exe (WSL). Works?

rsashank commented on tests/helper/test_helper.py:
```diff
@@ -603,28 +603,53 @@ def test_process_media_empty_url(
+        # NOTE: WSL always returns a non-zero exit code (1), so we do not test for it.
+    ],
+    ids=[
+        "Linux_os_user",
+        "Mac_os_user",
+        "WSL_os_user",
+        "Linux_os_error",
+        "Mac_os_error",
```
Got it!

rsashank commented:
Updated this PR, thanks for the feedback @neiljp!

neiljp commented on tests/helper/test_helper.py:
```diff
@@ -603,28 +603,49 @@ def test_process_media_empty_url(
+            1,
+            "open",
+            [
+                " The tool ",
+                ("footer_contrast", "open"),
+                " did not run successfully" ". Exited with ",
+                ("footer_contrast", "1"),
```
Reading this again, a reasonable PR after this would be to not output the error code, but provide more detailed output somewhere for followup later, a little like we do with the crash or threading errors. That, or try and diagnose further why the tool did not run properly.

We could also possibly provide a 'ran successfully' message in the footer for opening, but I've not looked at the code flow for that, and whether it would eg. overwrite other messages too rapidly?

rsashank replied:
Understood 👍 

Wouldn't diagnosing why the tool did not run properly be platform dependent though?

neiljp reviewed (commented):
@rsashank I'd like to get this merged to firm up the WSL behavior - while someone is testing it :)

Regards to the platform isolation, placing the remaining PLATFORM-dependent functionality in platform_code.py would be preferable and ease testing overall, and could be another refactor to explore in addition to the followup I mentioned in a comment.

rsashank committed:
platform_code: Add process_media_tool function.

The process_media_tool function determines and returns the media tool
used based on the OS.

rsashank commented:
@neiljp  I've updated this PR—thank you!

I added another commit to move the platform-based conditions in the helper function to platform_code.py.

Can you elaborate on this a bit?
> Reading this again, a reasonable PR after this would be to not output the error code, but provide more detailed output somewhere for followup later, a little like we do with the crash or threading errors. That, or try and diagnose further why the tool did not run properly.
